### PR TITLE
ScpRole API reference rework

### PR DESCRIPTION
**Pull request by louis1706**

### CODE REVIEW

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 1]
```diff
@@ -34,6 +32,16 @@ internal Scp096Role(Scp096GameRole baseRole)
...
+
+            if (!SubroutineModule.TryGetSubroutine(out Scp096RageCycleAbility scp096RageCycleAbility))
+                Log.Error("impossible to take Scp096RageCycleAbility in Scp096Role.cttor");
+            Scp096RageCycleAbility = scp096RageCycleAbility;
+            if (!SubroutineModule.TryGetSubroutine(out Scp096RageManager scp096RageManager))
+                Log.Error("impossible to take Scp096RageManager in Scp096Role.cttor");
+            Scp096RageManager = scp096RageManager;
+            if (!SubroutineModule.TryGetSubroutine(out Scp096TargetsTracker scp096TargetsTracker))
+                Log.Error("impossible to take Scp096TargetsTracker in Scp096Role.cttor");
+            Scp096TargetsTracker = scp096TargetsTracker;
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:44]:
```suggestion
            if (!SubroutineModule.TryGetSubroutine(out Scp096RageCycleAbility scp096RageCycleAbility))
                Log.Debug("Scp096RageCycleAbility subroutine not found in Scp096Role::ctor");
            
            Scp096RageCycleAbility = scp096RageCycleAbility;
            
            if (!SubroutineModule.TryGetSubroutine(out Scp096RageManager scp096RageManager))
                Log.Debug("Scp096RageManager subroutine not found in Scp096Role::ctor");

            Scp096RageManager = scp096RageManager;

            if (!SubroutineModule.TryGetSubroutine(out Scp096TargetsTracker scp096TargetsTracker))
                Log.Debug("Scp096TargetsTracker not found in Scp096Role::ctor");
                
            Scp096TargetsTracker = scp096TargetsTracker;
```

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 2]
```diff
@@ -51,58 +59,45 @@ internal Scp096Role(Scp096GameRole baseRole)
...
         /// </summary>
-        public bool IsEnraged => RageState == Scp096RageState.Enraged;
+        public Scp096RageManager Scp096RageManager { get; }
 
         /// <summary>
-        /// Gets a value indicating whether or not SCP-096 is currently docile.
+        /// Gets .
         /// </summary>
-        public bool IsDocile => !IsEnraged;
+        public Scp096TargetsTracker Scp096TargetsTracker { get; }
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:74]:
```suggestion
        public Scp096TargetsTracker TargetsTracker { get; }
```
Change the name of this property to the new one everytime it gets called above and below.

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 3]
```diff
@@ -51,58 +59,45 @@ internal Scp096Role(Scp096GameRole baseRole)
...
         /// </summary>
-        public bool CanReceiveTargets => SubroutineModule.TryGetSubroutine(out Scp096RageCycleAbility ability) && ability._targetsTracker.CanReceiveTargets;
+        public Scp096RageCycleAbility Scp096RageCycleAbility { get; }
 
         /// <summary>
-        /// Gets a value indicating whether or not SCP-096 is currently enraged.
+        /// Gets .
         /// </summary>
-        public bool IsEnraged => RageState == Scp096RageState.Enraged;
+        public Scp096RageManager Scp096RageManager { get; }
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:69]:
```suggestion
        public Scp096RageManager RageManager { get; }
```
Change the name of this property to the new one everytime it gets called above and below.

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 4]
```diff
@@ -51,58 +59,45 @@ internal Scp096Role(Scp096GameRole baseRole)
...
-        /// Gets a value indicating SCP-096's rage state.
-        /// </summary>
-        public Scp096RageState RageState => Internal.StateController.RageState;
-
-        /// <summary>
-        /// Gets a value indicating whether or not SCP-096 can receive targets.
+        /// Gets .
         /// </summary>
-        public bool CanReceiveTargets => SubroutineModule.TryGetSubroutine(out Scp096RageCycleAbility ability) && ability._targetsTracker.CanReceiveTargets;
+        public Scp096RageCycleAbility Scp096RageCycleAbility { get; }
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:64]:
```suggestion
        public Scp096RageCycleAbility RageCycleAbility { get; }
```
Change the name of this property to the new one everytime it gets called above and below.

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 5]
```diff
@@ -164,5 +150,42 @@ public IReadOnlyCollection<Player> Targets
...
+        /// .
+        /// </summary>
+        /// <param name="player">ee.</param>
+        /// <returns>e.</returns>
+        public bool IsObserved(Player player) => Scp096TargetsTracker.IsObservedBy(player.ReferenceHub);
+
+        /// <summary>
+        /// .
+        /// </summary>
+        public void ClearTarget() => Scp096TargetsTracker.ClearAllTargets();
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:189]:
```suggestion
        public void ClearTargets() => Scp096TargetsTracker.ClearAllTargets();
```

**Review by iopietro — CHANGES_REQUESTED** [review]

**Comment by louis1706** [comment]
i would like to keep all the Error Check for all the Beta in case of getting issue with them

**Comment by iopietro** [comment]
Ok, seems about right.

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 6]
```diff
@@ -143,26 +137,59 @@ public float EnragedTimeLeft
...
+
+        /// <summary>
+        /// .
+        /// </summary>
+        /// <param name="player">ee.</param>
+        /// <returns>e.</returns>
+        public bool HasTarget(Player player) => TargetsTracker.HasTarget(player.ReferenceHub);
+
+        /// <summary>
+        /// .
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:184]: Some comments are missing.

File: Exiled.API/Features/Roles/Scp096Role.cs [hunk 7]
```diff
@@ -143,26 +137,59 @@ public float EnragedTimeLeft
...
+        /// </summary>
+        /// <param name="player">ee.</param>
+        /// <returns>e.</returns>
+        public bool IsObserved(Player player) => TargetsTracker.IsObservedBy(player.ReferenceHub);
+
+        /// <summary>
+        /// .
+        /// </summary>
+        public void ClearTarget() => TargetsTracker.ClearAllTargets();
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp096Role.cs:193]:
```suggestion
        public void ClearTargets() => TargetsTracker.ClearAllTargets();
```

File: Exiled.API/Features/Roles/Scp106Role.cs [hunk 8]
```diff
@@ -43,90 +71,104 @@ internal Scp106Role(Scp106GameRole baseRole)
...
+        public bool UsePortal(Vector3 position, float cost = 0f)
         {
-            if (SubroutineModule.TryGetSubroutine(out Scp106StalkAbility ability))
-                ability.IsActive = true;
+            HuntersAtlasAbility._syncPos = position;
+            if (Vigor > cost)
+                return false;
+            HuntersAtlasAbility._estimatedCost = cost;
+            HuntersAtlasAbility.SetSubmerged(true);
+            return true;
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp106Role.cs:171]:
```suggestion
            HuntersAtlasAbility._syncPos = position;

            if (Vigor > cost)
                return false;

            HuntersAtlasAbility._estimatedCost = cost;
            HuntersAtlasAbility.SetSubmerged(true);

            return true;
```

File: Exiled.API/Features/Roles/Scp939Role.cs [hunk 9]
```diff
@@ -140,22 +181,22 @@ public float MimicryCooldown
...
-                ability.RemoveRecordingsOfPlayer(target.ReferenceHub);
-            }
+                return;
+            MimicryRecorder.RemoveRecordingsOfPlayer(target.ReferenceHub);
+            MimicryRecorder.SavedVoicesModified = true;
+        }
 
-            ability.SavedVoicesModified = true;
+        /// <summary>
+        /// Removes all recordings of player voices. Provide an optional target to remove all the recordings of a single player.
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp939Role.cs:193]: Where's the optional target?

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Roles/Scp049Role.cs [hunk 10]
```diff
@@ -39,57 +54,51 @@ internal Scp049Role(Scp049GameRole baseRole)
...
-        {
-            get
-            {
-                if (!IsRecalling || !SubroutineModule.TryGetSubroutine(out Scp049ResurrectAbility ability))
-                    return null;
-
-                return Player.Get(ability.CurRagdoll.Info.OwnerHub);
-            }
-        }
+        public Player RecallingPlayer => Player.Get(ResurrectAbility.CurRagdoll?.Info.OwnerHub);
```
[the remark this answers is not in the document]
Comment (louis1706) [reply Exiled.API/Features/Roles/Scp049Role.cs:85]: it's will prevent NRE no ?

File: Exiled.API/Features/Roles/Scp939Role.cs (hunk 9 above)
Comment (louis1706) [reply Exiled.API/Features/Roles/Scp939Role.cs:193]: miss to modify the comment

**Review by Thundermaker300 — APPROVED** [review]

**Pull request merged**